---
title: "Oxi_GWAS_rTASSEL"
author: "Chosen_Obih"
date: "2024-02-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Allocate maximum memory to rTASSEL, load rjava and set working directory. The memory allocation should always be done before loading rTASSEL
```{r}
options(java.parameters = c("-Xmx30g", "-Xms30g"))
library(rJava)
setwd("/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/")
```


```{r}
library(dplyr)
library(ggplot2)
library(magrittr)
library(rTASSEL)
library(tibble)
library(ggtree)
```

Create logging file. #This should be run once at the beginning of the analysis
```{r}
rTASSEL::startLogger(fullPath = NULL, fileName = "Feb_22")
```


Create path to input files
```{r}
# Load data
snp_file <- "/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/1_GBSCornell_IndicaPray_81347snps_271_accs.hmp.txt"
oxi_file <- "/home/chosen/Desktop/Rice_GWAS_project/data/oxi_data_CON.txt"
output_dir <- "/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/"
```

Load in genotype and phenotype data
```{r}
geno <- rTASSEL::readGenotypeTableFromPath(path = snp_file)

pheno <- rTASSEL::readPhenotypeFromPath(path = oxi_file)
```

Join genotype and phenotype data for data exploration
```{r}
geno_pheno <- readGenotypePhenotype(
  phenoPathDFOrObj = pheno,
  genoPathOrObj = geno
)
```

You can extract rTASSEL data to R data format
```{r}
geno_mat <- as.matrix(geno)
#geno_mat[1:5, 1:5] #view the first 5 rows and columns
```

Get summary tables
```{r}
geno_pheno %>% siteSummary() %>% head()
geno_pheno %>% taxaSummary() %>% head()
geno_pheno %>% taxaList() %>% head(20)
geno_pheno %>% positionList() %>% head()
```

PCA # This produces an S4 vector object. Not captured in tutorial!
```{r}
pcaRes <- rTASSEL::pca(geno_pheno)
#str(pcaRes)
```

Plot total variance for first 10 PCs
```{r}
nPCs <- 10 # set number of PCs to plot
pcaRes@results$Eigenvalues_Datum %>% # get data frame from the results S4 object, list object (pcaRes)
  head(n = nPCs) %>% # take the first 'n' PCs in the data frame
  ggplot() +
  aes(x = PC, y = proportion_of_total, group = 1) + #group is set to 1 because data is a "basic single vector data" not categories
  geom_line(color = "red") +
  geom_point(size = 3) +
  xlab("PC") +
  ylab("Proportion of total variance")
```

Load in meta data to plot PCA based on it.
```{r}
taxa_metadata <- read.csv("/home/chosen/Desktop/Rice_GWAS_project/data/meta_oxi_data.csv")
taxa_metadata <- taxa_metadata %>%
  mutate(Taxa = as.character(Taxa)) #change Taxa column from inteher to character.
```

Join the meta data dataframe to PC_Datum dataframe based on Taxa 
```{r}
pca_Datum <- pcaRes@results$PC_Datum
pca_Datum %>%
  left_join(taxa_metadata, by = "Taxa") %>%
  head()
```

Plot the new PCA
```{r}
pca_Datum %>%
  left_join(taxa_metadata, by = "Taxa") %>%
  ggplot() +
  aes(PC1, PC2, color = Treatment) +
  geom_point(size = 2)
```

Plot phylogentic tree. rTASSEL support two options UPMGA & neighbourhood joining. Auto = UPMGA
```{r}
phyloRes <- rTASSEL::createTree(geno_pheno)
```

Example of filterig the the phylogeny based on taxaList
```{r fig.height=40, fig.width=8}
geno_pheno %>%
  #filterGenotypeTableTaxa(txa = taxaList(.) %>% sample(size = 20)) %>%
  createTree() %>%
  plot()
```

Using ggtree library to modify the tree
```{r fig.height=5, fig.width=5}
geno_pheno %>%
  createTree() %>%
  ggtree(size = 1.5, layout = "circular", branch.length = "none") #%<+%
# meta_oxi_data +
# aes(color = Treatment)
```

Calculate sequence diversity
 - By default, this will generate a single set of diversity metrics for all markers
 - We can change this by activating a sliding window and step parameter which "scan" through sets of marker 
```{r}
seqRes <- seqDiversity(geno_pheno, slidingWindowAnalysis = TRUE, windowSize = 200, stepSize = 25)
seqRes$Diversity %>% head()
```


```{r fig.height=5, fig.width=12}
seqRes$Diversity %>%
  dplyr::filter(Chromosome == "1") %>%
  ggplot() +
  aes(x = StartChrPosition, y = TajimaD) +
  geom_line() +
  geom_point() +
  xlab("Position") +
  ylab("Tajima's D")
```

Filter genotype table sites
```{r}
geno_pheno_filtered <- rTASSEL::filterGenotypeTableSites(
    tasObj = geno_pheno,
    siteMinCount = 150,
    siteMinAlleleFreq = 0.05,
    siteMaxAlleleFreq = 1.0,
    siteRangeFilterType = "none"
)
```

Create a kinship matrix object
```{r}
kin <- rTASSEL::kinshipMatrix(tasObj = geno_pheno)
```

TAC GWAS GLM
```{r}
assocRes_GLM <- assocModelFitter(geno_pheno, formula = TAC ~ Treatment, fitMarkers = TRUE, maxP = 1)
```

TAC plotting GLM
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_GLM, trait = "TAC", threshold = 5, colors = c("#00AFBB", "#FC4E07"))
```

TAC GWAS MLM
```{r}
assocRes_MLM <- assocModelFitter(geno_pheno_filtered, formula = TAC ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

TAC plotting MLM
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM, trait = "TAC", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

DHAR GWAS GLM
```{r}
assocRes_GLM <- assocModelFitter(geno_pheno, formula = DHAR ~ Treatment, fitMarkers = TRUE, maxP = 1)
```

DHAR plotting GLM
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_GLM, trait = "DHAR", threshold = 4, colors = c("#00AFBB", "#FC4E07"))
```

DHAR GWAS MLM
```{r}
assocRes_MLM <- assocModelFitter(geno_pheno_filtered, formula = DHAR ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

DHAR plotting MLM
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM, trait = "DHAR", threshold = 4, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

MDHAR GWAS GLM
```{r}
assocRes_GLM <- assocModelFitter(geno_pheno, formula = MDHAR ~ Treatment, fitMarkers = TRUE, maxP = 1)
```

MDHAR plotting GLM
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_GLM, trait = "MDHAR", threshold = 5, colors = c("#00AFBB", "#FC4E07"))
```

MDHAR GWAS MLM
```{r}
assocRes_MLM <- assocModelFitter(geno_pheno_filtered, formula = MDHAR ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

MDHAR plotting MLM
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM, trait = "MDHAR", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```
