---
title: "Oxi_GWAS_statgen_DRO"
output: html_document
date: "2024-03-12"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Load package and set working directory
```{r}
library(statgenGWAS)
library(magrittr)
library(data.table)
setwd("/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/")
```

## read in files
```{r}
snp <- fread("/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/2.1_MARKERS.txt", header = TRUE)
oxi <- read.table("/home/chosen/Desktop/Rice_GWAS_project/data/oxi_data_statgen_DRO.txt", header = TRUE, sep = "\t")
marker <- read.csv("/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/2.1_mINFO.csv", header = TRUE)
```

## some data manipulation
```{r}
#Rename title of the first column and transpose the data
colnames(snp)[1] <- "SNP_name"
snp_t <- t(snp) %>% as.data.frame (header = TRUE)
df <- snp_t
df <- df[-1, ] # Remove the original first row
# Set the new first row (originally second) as column names
colnames(df) <- as.character(unlist(df[1, ]))
header_name <- as.character(unlist(snp_t[1, ]))
colnames(df) <- header_name
rownames(marker) <- marker[["SNP.names"]]
```

## create a gData object containing map and marker information.
```{r}
g_oxi_data <- createGData(geno = df, map = marker)
```

## split the oxi dataframe and create small data frame of each phenotype
```{r}
oxi_list <- split(x = oxi[c("genotype", "GOX", "HPR",	"POX",	"APX",	"GPX",	"CAT",	
                                  "GST",	"SOD",	"DHAR",	"MDHAR",	"AO",	"GR",	"Grxs",
                                  "Trxs",	"Prxs",	"Frxs",	"MDA",	"TAC",	"Poly",	"ProtOx")], 
                        f = oxi[["treatment"]])
```

## combine genotype and phenotype objects
```{r}
oxi_geno_pheno <- createGData(gData = g_oxi_data, pheno = oxi_list)
#summary(oxi_geno_pheno) #view the summary of the genotype-phenotype object
```

## Vizualize the position of the SNPs on the different chromosomes
```{r}
plot(oxi_geno_pheno)
```

## The SNP data is in the format of A/T G/C, it needs to be recoded to make it compactible with statgen GWAS function
```{r}
oxi_geno_pheno_recoded <- codeMarkers(gData = oxi_geno_pheno)
```

## single trait GWAS run
```{r}
DHAR_GWAS <- runSingleTraitGwas(gData = oxi_geno_pheno_recoded, traits = "DHAR")
#summary(DHAR_GWAS)
```

## plot QQ-plot to assess the quality of the GWAS model.
```{r}
plot(DHAR_GWAS, plotType = "qq", trait = "DHAR")
```


```{r fig.height=4, fig.width=10}
colors = c("#FC4E07", "#00AFBB", "#FC4E07", "#00AFBB","#FC4E07", "#00AFBB", "#FC4E07", "#00AFBB","#FC4E07", "#00AFBB", "#FC4E07", "#00AFBB")
plot(DHAR_GWAS, plotType = "manhattan", col = colors, trait = "DHAR", yThr = 4.5)
```

## DHAR GWAS including kingship matrix (vanRaden)
```{r}
DHAR_GWAS_kin_van <- runSingleTraitGwas(gData = oxi_geno_pheno_recoded, traits = "DHAR", GLSMethod = "multi", kinshipMethod = "vanRaden")
```

## plot DHAR GWAS kingship van
```{r fig.height=4, fig.width=10}
plot(DHAR_GWAS_kin_van, plotType = "manhattan", trait = "DHAR", yThr = 4.5, col = colors)
```

## DHAR GWAS including kingship matrix (astle)
```{r}
DHAR_GWAS_kin_ast <- runSingleTraitGwas(gData = oxi_geno_pheno_recoded, traits = "DHAR", GLSMethod = "multi", kinshipMethod = "astle")
```

## plot DHAR GWAS kinship astle
```{r fig.height=4, fig.width=10}
plot(DHAR_GWAS_kin_ast, plotType = "manhattan", trait = "DHAR", yThr = 4.5, col = colors)
```

## DHAR GWAS using fdr
```{r}
DHAR_GWAS_fdr <- runSingleTraitGwas(gData = oxi_geno_pheno_recoded, traits = "DHAR", thrType = "fdr")
```


```{r fig.height=4, fig.width=10}
plot(DHAR_GWAS_fdr, plotType = "manhattan", trait = "DHAR", yThr = 4.5, col = colors)
```

## DHAR GWAS using fdr and kingship (van)
```{r}
DHAR_GWAS_fdr_kin <- runSingleTraitGwas(gData = oxi_geno_pheno_recoded, traits = "DHAR", GLSMethod = "multi", kinshipMethod = "vanRaden", thrType = "fdr")
```

## plotting DHAR GWAS using fdr and kingship (van)
```{r fig.height=4, fig.width=10}
plot(DHAR_GWAS_fdr_kin, plotType = "manhattan", trait = "DHAR", yThr = 4.5, col = colors)
```


```{r}
plot(DHAR_GWAS_fdr, plotType = "qtl", yThr = 4.5)
```

