---
title: "Oxi_GWAS_rTASSEL"
author: "Chosen_Obih"
date: "2024-02-09"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## mere humans computational problem.
## I was having issues configuring rJava to allocate memory. I follow the steps below to fix it. Unfortunately, these steps have to be repeated every time Rstudio is restarted, but at least it works.
## Run each line in the console.
```{r}
remove.packages("rJava", lib="~/R/x86_64-pc-linux-gnu-library/4.3")

Sys.setenv(JAVA_HOME='/usr/lib/jvm/java-17-openjdk-amd64')

install.packages("rJava")
```

```{r}
library(rJava)
.jinit(parameters="-Xmx30g")
```


## Confirm that the specified memory was allocated.
```{r}
maxAllocRam <- .jcall(.jnew("java/lang/Runtime"), "J", "maxMemory") / 1e9 
message("You have ", round(maxAllocRam, 2), "GB memory available.", sep = "")
```

set working directory.
```{r}
setwd("/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/")
```


```{r}
library(dplyr)
library(ggplot2)
library(magrittr)
library(rTASSEL)
library(tibble)
library(ggtree)
```

Create logging file. #This should be run once at the beginning of the analysis
```{r}
rTASSEL::startLogger(fullPath = NULL, fileName = "Mar_20")
```


Create path to input files
```{r}
# Load data
#snp_file <- "/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/1_GBSCornell_IndicaPray_81347snps_271_accs.hmp.txt" #81k SNPs file

snp_file <- "/home/chosen/Desktop/Rice_GWAS_project/data/SNPs_files/GBSCornell_IndicaPray_Intermediate_339Samples.hmp.txt" #46k SNPs file
oxi_file <- "/home/chosen/Desktop/Rice_GWAS_project/data/clean/oxi_data_covariate.txt"
output_dir <- "/home/chosen/Desktop/Rice_GWAS_project/data/GENO_data/"
```

Load in genotype and phenotype data
```{r}
geno <- rTASSEL::readGenotypeTableFromPath(path = snp_file)

pheno <- rTASSEL::readPhenotypeFromPath(path = oxi_file)
```

Join genotype and phenotype data for data exploration
```{r}
geno_pheno <- readGenotypePhenotype(
  phenoPathDFOrObj = pheno,
  genoPathOrObj = geno
)
```

You can extract rTASSEL data to R data format
```{r}
geno_mat <- as.matrix(geno)
#geno_mat[1:5, 1:5] #view the first 5 rows and columns
```

Get summary tables
```{r}
geno_pheno %>% siteSummary() %>% head()
geno_pheno %>% taxaSummary() %>% head()
geno_pheno %>% taxaList() %>% head(20)
geno_pheno %>% positionList() %>% head()
```

PCA # This produces an S4 vector object. Not captured in tutorial!
```{r}
pcaRes <- rTASSEL::pca(geno_pheno)
#str(pcaRes)
```

Plot total variance for first 10 PCs
```{r}
nPCs <- 10 # set number of PCs to plot
pcaRes@results$Eigenvalues_Datum %>% # get data frame from the results S4 object, list object (pcaRes)
  head(n = nPCs) %>% # take the first 'n' PCs in the data frame
  ggplot() +
  aes(x = PC, y = proportion_of_total, group = 1) + #group is set to 1 because data is a "basic single vector data" not categories
  geom_line(color = "red") +
  geom_point(size = 3) +
  xlab("PC") +
  ylab("Proportion of total variance")
```

Load in meta data to plot PCA based on it.
```{r}
taxa_metadata <- read.csv("/home/chosen/Desktop/Rice_GWAS_project/data/meta_oxi_data.csv")
taxa_metadata <- taxa_metadata %>%
  mutate(Taxa = as.character(Taxa)) #change Taxa column from inteher to character.
```

Join the meta data dataframe to PC_Datum dataframe based on Taxa 
```{r}
pca_Datum <- pcaRes@results$PC_Datum
pca_Datum %>%
  left_join(taxa_metadata, by = "Taxa") %>%
  head()
```

Plot the new PCA
```{r}
pca_Datum %>%
  left_join(taxa_metadata, by = "Taxa") %>%
  ggplot() +
  aes(PC1, PC2, color = Treatment) +
  geom_point(size = 2)
```

Plot phylogentic tree. rTASSEL support two options UPMGA & neighbourhood joining. Auto = UPMGA
```{r}
phyloRes <- rTASSEL::createTree(geno_pheno)
```

Example of filterig the the phylogeny based on taxaList
```{r fig.height=40, fig.width=8}
geno_pheno %>%
  #filterGenotypeTableTaxa(txa = taxaList(.) %>% sample(size = 20)) %>%
  createTree() %>%
  plot()
```

Using ggtree library to modify the tree
```{r fig.height=5, fig.width=5}
geno_pheno %>%
  createTree() %>%
  ggtree(size = 1.5, layout = "circular", branch.length = "none") #%<+%
# meta_oxi_data +
# aes(color = Treatment)
```

Calculate sequence diversity
 - By default, this will generate a single set of diversity metrics for all markers
 - We can change this by activating a sliding window and step parameter which "scan" through sets of marker 
```{r}
seqRes <- seqDiversity(geno_pheno, slidingWindowAnalysis = TRUE, windowSize = 200, stepSize = 25)
seqRes$Diversity %>% head()
```


```{r fig.height=5, fig.width=12}
seqRes$Diversity %>%
  dplyr::filter(Chromosome == "1") %>%
  ggplot() +
  aes(x = StartChrPosition, y = TajimaD) +
  geom_line() +
  geom_point() +
  xlab("Position") +
  ylab("Tajima's D")
```

Filter genotype table sites
```{r}
geno_pheno_filtered <- rTASSEL::filterGenotypeTableSites(
    tasObj = geno_pheno,
    siteMinCount = 150,
    siteMinAlleleFreq = 0.05,
    siteMaxAlleleFreq = 1.0,
    siteRangeFilterType = "none"
)
```

Create a kinship matrix object
```{r}
kin <- rTASSEL::kinshipMatrix(tasObj = geno_pheno_filtered, method = "Centered_IBS", maxAlleles = 6)
```

#Glycolate oxidase (GOX) - photorespiration enzyme
```{r}
assocRes_MLM_GOX <- assocModelFitter(geno_pheno_filtered, formula = GOX ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#GOX plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_GOX, trait = "GOX", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Hydroxpyruvate reductase (HPR) - photorespiration enzyme
```{r}
assocRes_MLM_HPR <- assocModelFitter(geno_pheno_filtered, formula = HPR ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#HPR plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_HPR, trait = "HPR", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Peroxidase (POX) - Antioxidant enzyme
```{r}
assocRes_MLM_POX <- assocModelFitter(geno_pheno_filtered, formula = POX ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#POX plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_POX, trait = "POX", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Ascorbate peroxidase (APX) - Antioxidant enzyme
```{r}
assocRes_MLM_APX <- assocModelFitter(geno_pheno_filtered, formula = APX ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#APX plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_APX, trait = "APX", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Glutathione peroxidase (GPX) - Antioxidant enzyme
```{r}
assocRes_MLM_GPX <- assocModelFitter(geno_pheno_filtered, formula = GPX ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#GPX plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_GPX, trait = "GPX", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Catalse (CAT) - Antioxidant enzyme
```{r}
assocRes_MLM_CAT <- assocModelFitter(geno_pheno_filtered, formula = CAT ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#CAT plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_CAT, trait = "CAT", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Glutathione S-transferase (GST) - Antioxidant enzyme
```{r}
assocRes_MLM_GST <- assocModelFitter(geno_pheno_filtered, formula = GST ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#GST plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_GST, trait = "GST", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Superoxide dismutase (SOD) - Antioxidant enzyme
```{r}
assocRes_MLM_SOD <- assocModelFitter(geno_pheno_filtered, formula = SOD ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#SOD plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_SOD, trait = "SOD", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Ascorbate oxidase (AO) - Antioxidant enzyme
```{r}
assocRes_MLM_AO <- assocModelFitter(geno_pheno_filtered, formula = AO ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#AO plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_AO, trait = "AO", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Glutathione reductase (GR) - Antioxidant enzyme
```{r}
assocRes_MLM_GR <- assocModelFitter(geno_pheno_filtered, formula = GR ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#GR plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_GR, trait = "GR", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Glutaredoxins (Grxs) - Antioxidant enzyme
```{r}
assocRes_MLM_Grxs <- assocModelFitter(geno_pheno_filtered, formula = Grxs ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#Grxs plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_Grxs, trait = "Grxs", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Thioredoxins (Trxs) - Antioxidant enzyme
```{r}
assocRes_MLM_Trxs <- assocModelFitter(geno_pheno_filtered, formula = Trxs ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#Trxs plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_Trxs, trait = "Trxs", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Peroxiredoxins (Prxs) - Antioxidant enzyme
```{r}
assocRes_MLM_Prxs <- assocModelFitter(geno_pheno_filtered, formula = Prxs ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#Prxs plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_Prxs, trait = "Prxs", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Ferredoxins (Frxs) - Antioxidant enzyme
```{r}
assocRes_MLM_Frxs <- assocModelFitter(geno_pheno_filtered, formula = Frxs ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#Frxs plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_Frxs, trait = "Frxs", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Malondialdehyde (MDA) - Oxidative stress marker 
```{r}
assocRes_MLM_MDA <- assocModelFitter(geno_pheno_filtered, formula = MDA ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#MDA plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_MDA, trait = "MDA", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Total polyphenol content (Poly) - molecular antioxidant
```{r}
assocRes_MLM_Poly <- assocModelFitter(geno_pheno_filtered, formula = Poly ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#Poly plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_Poly, trait = "Poly", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Protein oxidation (ProtOx) - Oxidative stress marker 
```{r}
assocRes_MLM_ProtOx <- assocModelFitter(geno_pheno_filtered, formula = ProtOx ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#ProtOx plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_ProtOx, trait = "ProtOx", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Dehydroascorbate redutase (DHAR) - Antioxidant enzyme
```{r}
assocRes_MLM_DHAR <- assocModelFitter(geno_pheno_filtered, formula = DHAR ~ Treatment, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#Dehydroascorbate redutase (DHAR) - Antioxidant enzyme - GWAS with PCs as covariate
```{r}
assocRes_MLM_DHAR_PC <- assocModelFitter(geno_pheno_filtered, formula = DHAR ~ Treatment + PC1, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#DHAR plotting - PC 1-5
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_DHAR_PC, trait = "DHAR", threshold = 4.5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#DHAR plotting PC 1-3
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_DHAR, trait = "DHAR", threshold = 4.5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Monodehydroascrobate reductas (MDHAR) - Antioxidant enzyme
```{r}
assocRes_MLM_MDHAR <- assocModelFitter(geno_pheno_filtered, formula = MDHAR ~ Treatment + PC1 + PC2 + PC3, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#MDHAR plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_MDHAR, trait = "MDHAR", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

#Total Antioxidant Capacity (TAC) - molecular antioxidant
```{r}
assocRes_MLM_TAC_PC <- assocModelFitter(geno_pheno_filtered, formula = TAC ~ Treatment + PC1, fitMarkers = TRUE, kinship = kin, maxP = 1)
```

#TAC plotting
```{r fig.height=4, fig.width=12}
plotManhattan(assocRes_MLM_TAC_PC, trait = "TAC", threshold = 5, interactive = FALSE, colors = c("#00AFBB", "#FC4E07"))
```

Extracting TAC GWAS result
```{r}
result_TAC_PC <- assocRes_MLM_TAC_PC@results$MLM_Stats
write.csv(result_TAC_PC, "DRO_TAC_PC_MLM_stats.csv", row.names = FALSE)
```

Extracting MDHAR GWAS result
```{r}
result_MDHAR <- assocRes_MLM_MDHAR@results$MLM_Stats
write.csv(result_MDHAR, "DRO_MDHAR_MLM_stats.csv", row.names = FALSE)
```

LD plotting for MDHAR
```{r fig.height=8, fig.width=8}
# Filter genotype table by position
tasGenoPhenoFilt <- filterGenotypeTableSites(
    tasObj              = geno_pheno_filtered,
    siteRangeFilterType = "position",
    startPos            = 21802147,
    endPos              = 21935058,
    startChr            = 6,
    endChr              = 6
)

# Generate and visualize LD
myLD <- ldPlot(
    tasObj  = tasGenoPhenoFilt,
    ldType  = "All",
    plotVal = "r2",
    verbose = FALSE
)

myLD
```

Extracting DHAR GWAS result
```{r}
result_DHAR <- assocRes_MLM_DHAR@results$MLM_Stats
write.csv(result_DHAR, "DRO_DHAR_PC1_3_MLM_stats.csv", row.names = FALSE)
```

Extract PC
```{r}
write.csv(pca_Datum, "pca_oxi_markers.csv", row.names = FALSE)
```


LD plotting for DHAR
```{r fig.height=8, fig.width=8}
# Filter genotype table by position
tasGenoPhenoFilt <- filterGenotypeTableSites(
    tasObj              = geno_pheno_filtered,
    siteRangeFilterType = "position",
    startPos            = 21883335,
    endPos              = 21935058,
    startChr            = 6,
    endChr              = 6
)

# Generate and visualize LD
myLD_DHAR <- ldPlot(
    tasObj  = tasGenoPhenoFilt,
    ldType  = "All",
    plotVal = "r2",
    verbose = FALSE
)

myLD_DHAR
```

Extracting DHAR GWAS result
```{r}
result_TAC <- assocRes_MLM_TAC@results$MLM_Stats
write.csv(result_TAC, "DRO_TAC_PC1_3_MLM_stats.csv", row.names = FALSE)
```

LD plotting for TAC - chr 6
```{r fig.height=9, fig.width=13}
# Filter genotype table by position
tasGenoPhenoFilt <- filterGenotypeTableSites(
    tasObj              = geno_pheno_filtered,
    siteRangeFilterType = "position",
    startPos            = 21800143,
    endPos              = 21996438,
    startChr            = 6,
    endChr              = 6
)

# Generate and visualize LD
myLD_TAC <- ldPlot(
    tasObj  = tasGenoPhenoFilt,
    ldType  = "All",
    plotVal = "r2",
    verbose = FALSE
)

myLD_TAC
```

```{r}
function (tasGenoPhenoFilt, ldType = "All", windowSize = NULL, 
  hetCalls = c("missing", "ignore", "third"), plotVal = "r2", verbose = TRUE) 
{
  angle <- 135
  text_scale_buffer <- 1
  upper_margin_padding <- 3
  plotVal <- match.arg(plotVal)
  ldType <- match.arg(ldType)
  ldDF <- linkageDiseq(tasObj = tasObj, ldType = ldType, windowSize = windowSize, 
    hetCalls = hetCalls, verbose = verbose)
  if (plotVal == "r2") 
    plotVal <- "R^2"
  ldDF$coord1 <- paste0(ldDF$Locus1, "_", ldDF$Position1)
  ldDF$coord2 <- paste0(ldDF$Locus2, "_", ldDF$Position2)
  ldSub <- ldDF[, c("coord1", "coord2", plotVal)]
  ldSub <- as.data.frame(ldSub)
  ldSubRot <- ldCellRotater(ldSub, angle)
  ids <- unique(c(ldSub$coord1, ldSub$coord2))
  ids <- ids[order(ids)]
  id_coord <- list(x = seq(1.5, 1.5 + length(ids) - 1, 1), 
    y = seq(0.5, 0.5 + length(ids) - 1, 1))
  id_coord <- rotate(id_coord$x, id_coord$y, angle)
  legend_lab <- switch(EXPR = plotVal, `R^2` = bquote(italic(r)^2), 
    DPrime = bquote(italic(D) * "'"), pDiseq = bquote(italic(p) * 
      "-value"))
  cell_plot <- ggplot2::ggplot(data = ldSubRot) + ggplot2::aes(x = .data$x, 
    y = .data$y, fill = .data$val, group = .data$group) + 
    ggplot2::geom_polygon(color = "white", linewidth = 1) + 
    ggplot2::annotate(geom = "text", x = id_coord$x, y = id_coord$y + 
      (abs(id_coord$y) * text_scale_buffer), label = ids, 
      angle = 90) + ggplot2::ylim(min(ldSubRot$y), upper_margin_padding) + 
    ggplot2::scale_x_reverse() + ggplot2::scale_fill_continuous(name = legend_lab, 
    type = "viridis") + ggplot2::coord_fixed() + ggplot2::theme_void() + 
    ggplot2::theme(legend.position = "bottom")
  return(cell_plot)
}
```

```{r}
print(cell_plot)
```


LD plotting for TAC - chr 1
```{r fig.height=12, fig.width=20}
# Filter genotype table by position
tasGenoPhenoFilt <- filterGenotypeTableSites(
    tasObj              = geno_pheno_filtered,
    siteRangeFilterType = "position",
    startPos            = 3695146,
    endPos              = 3938466,
    startChr            = 1,
    endChr              = 1
)

# Generate and visualize LD
myLD_TAC_1 <- ldPlot(
    tasObj  = tasGenoPhenoFilt,
    ldType  = "All",
    plotVal = "r2",
    verbose = FALSE
)

myLD_TAC_1
```


```{r}

```

